### PR TITLE
Capture initial working thoughts about claims language.

### PR DESCRIPTION
**Pull request by bbockelm**
Define some prototype claim language and take a first stab at the corresponding URIs and examples.

### CODE REVIEW

**Comment by jbasney** [comment]
I think we should be specifying the scope of the access using the standard OAuth [scope](https://tools.ietf.org/html/rfc6749#section-3.3) claim, not defining new claims for access control. Then we get all the scope-based functionality of OAuth, e.g., requesting different scopes at different times.

Rather than a vo claim, I'd prefer an isMemberOf claim, following the standard set by the [REFEDS OIDCRE Group](https://wiki.refeds.org/display/GROUPS/Mapping+SAML+attributes+to+OIDC+Claims).

I'd like to conform to the [OIDC ID Token](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) format, e.g., require a sub claim.

**Comment by bbockelm** [comment]
@jbasney -
* I'm not sure I understand your comment about the scopes?  Can you give an example?  From the OAuth link you gave (which, to be honest, I don't have deep enough expertise on OAuth yet...), it seems that you are thinking of renaming `op` in the PR to `scope`?  That is fine by me, but your comment seems to indicate there's a deeper semantic here...
* I'm happy to migrate `vo` to `isMemberOf` (honestly, I don't feel strongly about any of the attribute names at this point).  However:
   * I noticed that, in the link you give, `isMemberOf` is listed as an orphaned attribute because of its wide pre-existing use.
   * `vo` makes more sense when you see the (currently unposted) document about verifying tokens.  Distributed verification implies some mapping from signing key to trust roots -- I was hoping to use the `vo` value as a globally unique hash key to determine a set of public signing keys.
   * Hence, the attribute becomes part of the _global_ namespace, not the VO's namespace.
   * If OIDC has an equivalent, I'd be happy to use that.  Not entirely sure `isMemberOf` is the right one, however...

**Comment by bbockelm** [comment]
In terms of the OIDC ID Token format, even if this isn't meant to be an ID token, perhaps there are still claims we can reuse?  Reading through:
   * `iss`: Required in OIDC, optional here.  I'm OK with making this required.
   * `sub`: Required in OIDC, optional here.  Given we want to emphasize a capabilities-based ecosystem instead of an identity-based, I'd really like to keep this optional.
   * `aud`: Is this close enough to the SciTokens-specific `svc` claim that we can remove `svc` and use this instead?
   * `exp`: Required by both.  We can copy over the more formal language though?
   * `iat`: Not defined by SciTokens.  Required by OIDC.  I think it should at least be optional.
   * `auth_time`: Not sure if it has much value including it here in addition to the ID token.
   * `nonce`: Guard against replay attacks.  Seems reasonable to add?
   * `acr`/`amr`: Seems reasonable to include this as optional.
   * `azp`: Again, seems reasonable to include this as optional.

A lot of the optional items are only optional because I'm a bit paranoid about the length of the serialized token and would be happy to include them.

Any of these in particular you'd like reused?

**Comment by jbasney** [comment]
I advocate for using [scope](https://tools.ietf.org/html/rfc6749#section-3.3) because it is a standard OAuth parameter with defined semantics around requesting, granting, matching, narrowing, defaults, etc. 

Example request:
```
  GET /authorize?response_type=code&client_id=6e8fdae3459dac6c&redirect_uri=https://localhost/callback&scope=https://ligo.org/scopes/ligo.prod.o1.cbc.bbh.pycbcoffline+https://ligo.org/scopes/ligo.prod.o1.cbc.jobscratch.234987
```

Example [OAuth 2.0 Token Introspection](https://tools.ietf.org/html/rfc7662) response to determine the authorized scopes:
```
     HTTP/1.1 200 OK
     Content-Type: application/json

     {
      "active": true,
      "client_id": "l238j323ds-23ij4",
      "username": "jdoe",
      "scope": "https://ligo.org/scopes/ligo.prod.o1.cbc.bbh.pycbcoffline https://ligo.org/scopes/ligo.prod.o1.cbc.jobscratch.234987",
      "sub": "Z5O3upPC88QrAjx00dis",
      "aud": "https://protected.example.net/resource",
      "iss": "https://server.example.com/",
      "exp": 1419356238,
      "iat": 1419350238,
     }
```

File: technical_docs/Claims.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+>   A producer and consumer of a JWT MAY agree to use Claim Names that
+>   are Private Names: names that are not Registered Claim Names
+>   (Section 4.1) or Public Claim Names (Section 4.2).  Unlike Public
+>   Claim Names, Private Claim Names are subject to collision and should
+>   be used with caution.
+
+In the long-term, we hope to have our claim names registered with IANA.  Until then, we will
+utilize the URI form.  A SciTokens validator MUST accept either URI or non-URI form.
+
+* *vo* (Virtual Organization): https://scitokens.org/v1/vo. The virtual organization issuing the token.  This MUST be globally unique for the community utilizing the SciToken.  It MUST be utilized to determine the appropriate trust roots for token verification.  If present in a token, it MUST also be present in the header.  A token chain MUST have the VO specified in the base token.
```
Comment (jbasney) [inline technical_docs/Claims.md]: I'm confused about the relationship between *vo* and *iss*. I expect the JWT *iss* value to determine the cryptographic keys for verification, not the *vo* value. Quoting [draft-sheffer-oauth-jwt-bcp-01](https://tools.ietf.org/html/draft-sheffer-oauth-jwt-bcp-01), "When a JWT contains an "iss" (issuer) claim, the application MUST validate that the cryptographic keys used for the cryptographic operations in the JWT belong to the issuer.  If they do not, the application MUST reject the JWT."

File: technical_docs/Claims.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+* *op* (Authorized Operation): https://scitokens.org/v1/op.  A list of operations the bearer is allowed to perform.  Known operations are:
+
+   * `read`: Read data.
+   * `write`: Write data.
+   * `queue`: Submit a task or a job to a queueing service.
+   * `execute`: Immediately launch or execute a task.
+
+   The operation definitions are currently kept open-ended and intended to be interpreted by the specific user community.  The `op` claim is OPTIONAL.
+
+* *svc* (Service): https://scitokens.org/v1/svc.  A service URI the SciToken is authorized to access.  For example, if the VO has write access to several storage services, this claim may be utilized to limit a token to a single endpoint.  The `svc` claim is OPTIONAL.
```
Comment (jbasney) [inline technical_docs/Claims.md]: I think this is the JWT [aud](https://tools.ietf.org/html/rfc7519#section-4.1.3) claim?

File: technical_docs/Claims.md (hunk 1 above)
Comment (bbockelm) [inline technical_docs/Claims.md]: I was thinking of having this be a field for internal information from the VO.  I agree it makes more sense to have it use my standard OAuth-style.  Updating...

File: technical_docs/Claims.md (hunk 2 above)
Comment (bbockelm) [inline technical_docs/Claims.md]: The part of the `aud` claim I struggled with was:
```
   If the principal
   processing the claim does not identify itself with a value in the
   "aud" claim when this claim is present, then the JWT MUST be
   rejected.
```

I was worried whether there should be any distinction between a set of services and the audience (this worry was probably more obvious when I called this the `site`, which is a bit more general).

Given we want to lean on the side of keeping with OAuth, let's switch to the `aud` claim and revisit as necessary.

**Comment by bbockelm** [comment]
@jbasney - sorry for the delayed response.  Pushed some fixes.  See anything else that needs tackling?

**Comment by bbockelm** [comment]
Chatted with @jbasney over Skype.  Going ahead and merging for now; will do any additional work in follow-up PRs.

**Pull request merged**